### PR TITLE
gossip/processor: use PTrit hashing for improved performance

### PR DESCRIPTION
**Pull request by th0br0**

### CODE REVIEW

File: gossip/components/processor.c [hunk 1]
```diff
@@ -259,30 +275,63 @@ static void *processor_routine(processor_t *const processor) {
...
-    if (process_packet(processor, &packet) != RC_OK) {
-      log_warning(PROCESSOR_LOGGER_ID, "Processing packet failed\n");
+    if(packet_cnt == 0) {
+      continue;
+    }
+
+    ptrit_curl_init(curl, CURL_P_81);
+    memset(txs_acc, 0, NUM_TRITS_SERIALIZED_TRANSACTION * sizeof(ptrit_t));
+    memset(hashes_acc, 0, HASH_LENGTH_TRIT * sizeof(ptrit_t));
+    memset(&flex_hash, FLEX_TRIT_NULL_VALUE, sizeof(flex_hash));
```
Comment (thibault-martinez) [inline gossip/components/processor.c]:
```suggestion
    memset(flex_hash, FLEX_TRIT_NULL_VALUE, sizeof(flex_hash));
```

File: gossip/components/processor.c [hunk 2]
```diff
@@ -259,30 +275,63 @@ static void *processor_routine(processor_t *const processor) {
...
+
+    for(j = 0; j < packet_cnt; j++) {
+      bytes_to_trits(&packets[j].content, PACKET_SIZE, tx, NUM_TRITS_SERIALIZED_TRANSACTION);
```
Comment (thibault-martinez) [inline gossip/components/processor.c]:
```suggestion
      bytes_to_trits(packets[j].content, PACKET_SIZE, tx, NUM_TRITS_SERIALIZED_TRANSACTION);
```

File: gossip/components/processor.c [hunk 3]
```diff
@@ -259,30 +275,63 @@ static void *processor_routine(processor_t *const processor) {
...
+      trits_to_ptrits(tx, txs_acc, j, NUM_TRITS_SERIALIZED_TRANSACTION);
+    }
+
+    ptrit_curl_absorb(curl, txs_acc, NUM_TRITS_SERIALIZED_TRANSACTION);
+    ptrit_curl_squeeze(curl, hashes_acc, HASH_LENGTH_TRIT);
+
+    for(j = 0; j < packet_cnt; j++) {
+      ptrits_to_trits(hashes_acc, hash, j, HASH_LENGTH_TRIT);
+      flex_trits_from_trits(&flex_hash, HASH_LENGTH_TRIT, hash, HASH_LENGTH_TRIT, HASH_LENGTH_TRIT);
```
Comment (thibault-martinez) [inline gossip/components/processor.c]:
```suggestion
      flex_trits_from_trits(flex_hash, HASH_LENGTH_TRIT, hash, HASH_LENGTH_TRIT, HASH_LENGTH_TRIT);
```

File: gossip/components/processor.c [hunk 4]
```diff
@@ -259,30 +275,63 @@ static void *processor_routine(processor_t *const processor) {
...
+      
+      if (process_packet(processor, &packets[j], &flex_hash) != RC_OK) {
```
Comment (thibault-martinez) [inline gossip/components/processor.c]:
```suggestion
      if (process_packet(processor, &packets[j], flex_hash) != RC_OK) {
```

File: common/curl-p/ptrit.c [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+void ptrit_curl_init(PCurl *const ctx, CurlType type) {
+  ptrit_curl_reset(ctx);
+  ctx->type = type;
+}
+
+void ptrit_curl_absorb(PCurl *const ctx, ptrit_t const *const trits,
+                       size_t length) {
+  size_t numChunks =
+      length / HASH_LENGTH_TRIT + ((length % HASH_LENGTH_TRIT) ? 1 : 0);
```
Comment (jserv) [inline common/curl-p/ptrit.c:36]: How about change from `(length % HASH_LENGTH_TRIT) ? 1 : 0` to `!!(length % HASH_LENGTH_TRIT)`?

File: common/model/transaction.c [hunk 6]
```diff
@@ -24,7 +24,8 @@
 // Fills up an existing transaction with the serialized data in bytes
 // Return non 0 on success
 size_t transaction_deserialize_trits(iota_transaction_t transaction,
-                                     const flex_trit_t *trits) {
+                                     const flex_trit_t *trits,
+                                     bool computeHash) {
```
Comment (thibault-martinez) [inline common/model/transaction.c]: snake_case

File: consensus/test_utils/tangle.c [hunk 7]
```diff
@@ -38,14 +38,14 @@ retcode_t tangle_cleanup(tangle_t *const tangle, char *test_db_path) {
 }
 
 void transactions_deserialize(tryte_t const *const *const transactions_trytes,
-                              iota_transaction_t *txs,
-                              size_t num_transactions) {
+                              iota_transaction_t *txs, size_t num_transactions,
+                              bool computeHash) {
```
Comment (thibault-martinez) [inline consensus/test_utils/tangle.c]: snake_case

File: consensus/test_utils/tangle.h [hunk 8]
```diff
@@ -23,7 +23,8 @@ retcode_t tangle_setup(tangle_t *const tangle,
 retcode_t tangle_cleanup(tangle_t *const tangle, char *test_db_path);
 
 void transactions_deserialize(tryte_t const *const *const transactions_trytes,
-                              iota_transaction_t *txs, size_t num_transactions);
+                              iota_transaction_t *txs, size_t num_transactions,
+                              bool computeHash);
```
Comment (thibault-martinez) [inline consensus/test_utils/tangle.h]: snake_case

File: common/curl-p/ptrit.c [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    length = length < HASH_LENGTH_TRIT ? 0 : length - HASH_LENGTH_TRIT;
+  }
+}
+
+void ptrit_transform(PCurl *const ctx) {
+  PCurl s = {0};
+  size_t round = 0;
+  for (; round < ctx->type; ++round) {
+    if (round & 1)
+      ptrit_sbox(ctx->state, &s.state);
```
Comment (thibault-martinez) [inline common/curl-p/ptrit.c:65]: `&s.state` will produce a warning and I'm not sure it'll work properly on all archs

File: common/curl-p/ptrit.c [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    else
+      ptrit_sbox(&s.state, ctx->state);
```
Comment (thibault-martinez) [inline common/curl-p/ptrit.c:67]: Ditto

File: common/model/transaction.h [hunk 11]
```diff
@@ -405,7 +405,8 @@ extern uint8_t transaction_weight_magnitude(
 // Creates and returns a new transaction without data
 iota_transaction_t transaction_new(void);
 // Creates and returns a new transaction from serialized data - NULL if failed
-iota_transaction_t transaction_deserialize(const flex_trit_t *trits);
+iota_transaction_t transaction_deserialize(const flex_trit_t *trits,
+                                           bool computeHash);
```
Comment (thibault-martinez) [inline common/model/transaction.h]: snake_case

File: gossip/components/processor.c [hunk 12]
```diff
@@ -62,13 +65,14 @@ static retcode_t process_transaction_bytes(processor_t const *const processor,
...
-  if (transaction_deserialize_from_trits(&transaction,
-                                         transaction_flex_trits) !=
+  if (transaction_deserialize_from_trits(&transaction, transaction_flex_trits,
+                                         false) !=
       NUM_TRITS_SERIALIZED_TRANSACTION) {
     log_warning(PROCESSOR_LOGGER_ID, "Deserializing transaction failed\n");
     ret = RC_PROCESSOR_INVALID_TRANSACTION;
     goto failure;
   }
+  memcpy(transaction_hash(&transaction), curl_hash, FLEX_TRIT_SIZE_243);
```
Comment (thibault-martinez) [inline gossip/components/processor.c]: Use `transaction_set_hash` instead

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**